### PR TITLE
feature/CreatePostgresDB

### PR DESCRIPTION
**Pull request by Tony-Thawatchai**
Description
Issue Link: [https://github.com/CodeForBc/airbnb-regulation/issues/30]

Dockerfile for creating postgres DB base on the data structure. And Github action workflow - revised version, removing postgres file

Type of Change
Please delete options that are not relevant.

[x ] New feature (non-breaking change which adds functionality)
How Has This Been Tested?
no

Checklist:
Before you submit your pull request, please make sure you have completed the following:

[ x] I have read the CONTRIBUTING document.
[ x] I have checked that my code adheres to the code style of this project.
[ x] I have performed a self-review of my own code.
[ x] I have commented my code, particularly in hard-to-understand areas.
 I have made corresponding changes to the documentation.
 My changes generate no new warnings.
 I have added tests that prove my fix is effective or that my feature works.
 New and existing unit tests pass locally with my changes.
 Any dependent changes have been merged and published in downstream modules.
Screenshots (if applicable)
Additional Notes

### CODE REVIEW

File: airbnb_project/database/.github\workflows/deploy-image.yml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v3
+
+      - name: Log in to the Container registry
+        uses: docker/login-action@f054a8b539a109f9f41c372932f1ae047eff08c9
+        with:
+          registry: ${{ env.REGISTRY }}
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
```
Comment (nam-m) [inline airbnb_project/database/.github\workflows/deploy-image.yml:25]: this means we need to set up github token? how do we do that?

File: airbnb_project/database/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# Use the official Docker Hub Postgres image
+FROM postgres
+
+# Set the working directory
+WORKDIR /var/lib/postgresql/data
```
Comment (nam-m) [inline airbnb_project/database/Dockerfile:5]: why you selected this directory?

File: airbnb_project/database/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+# Set the required environment variable
+ENV POSTGRES_PASSWORD=@irBnbRegulation
```
Comment (nam-m) [inline airbnb_project/database/Dockerfile:8]: probably should not expose env setting here

File: airbnb_project/database/airbnb_db_creator.sql [hunk 4]
```diff
@@ -0,0 +1,45 @@
+CREATE TABLE IF NOT EXISTS listing (
+    airbnb_listing_id VARCHAR(20) PRIMARY KEY,
```
Comment (gmartinsnull) [inline airbnb_project/database/airbnb_db_creator.sql:2]: I'd probably increase this to 50. They may not be that long today but could be the case they get longer as time goes on.

File: airbnb_project/database/airbnb_db_creator.sql [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    baths VARCHAR(10),
+    beds VARCHAR(10),
```
Comment (gmartinsnull) [inline airbnb_project/database/airbnb_db_creator.sql:4]: Ok a couple of thins here:
- maybe confirm with @immangat  but I think `baths`, `beds` and `person_capacity` should be `INTEGER`
- lat/lng could probably be `DECIMAL(8,6), DECIMAL(9,6)`, respectively
- `title` can be fairly long too so I'd suggest 200

File: airbnb_project/database/Dockerfile (hunk 2 above)
Comment (gmartinsnull) [reply airbnb_project/database/Dockerfile:5]: So this is for the case where we wanna compose the "app", say on a EC2 instance, that will be the default directory to which the app will be installed in. This could be anything. Matter of fact, this could simply be `/app` but I'm assuming it goes deeper in hierarchy because it's a db and not an application.

File: airbnb_project/database/Dockerfile (hunk 3 above)
Comment (gmartinsnull) [reply airbnb_project/database/Dockerfile:8]: Agreed. This should be probably done on the server level and not here.

File: airbnb_project/database/.github\workflows/deploy-image.yml [hunk 6]
```diff
@@ -0,0 +1,42 @@
+name: Create and publish a Docker image
+
+on: push
+
+env:
+  REGISTRY: ghcr.io
```
[the remark this answers is not in the document]
Comment (Tony-Thawatchai) [reply airbnb_project/database/.github\workflows/deploy-image.yml:6]: @gmartinsnull  I would like to discuss this topic with you whether we need the Github packages. I think we can answer this question if we have the decision made about how we going to deploy our app. Please feel free to advise me on this since I'm very new to Docker and CI/CD process.

**Comment by Tony-Thawatchai** [comment]
@nam-m  @gmartinsnull @immangat  I updated this PR to combine all services in one container according to our meeting agreement. Please help me review it and guide me on how to improve it. 

I have one concern: When I tested, I ran "docker-compose build," and it responded with a warning that the value in the .env file was not defined even though it was there. I tried finding a way and found that it works if I put the .env file in both the root folder and the ./database folder. I'm not sure if this is the right way or not.

File: airbnb_project/.DS_Store
Comment (gmartinsnull) [inline airbnb_project/.DS_Store:1]: so we should have a `.gitignore` and add this file to it so we refrain from having unnecessary/unused files on our repo

**Comment by gmartinsnull** [comment]
> @nam-m @gmartinsnull @immangat I updated this PR to combine all services in one container according to our meeting agreement. Please help me review it and guide me on how to improve it.
> 
> I have one concern: When I tested, I ran "docker-compose build," and it responded with a warning that the value in the .env file was not defined even though it was there. I tried finding a way and found that it works if I put the .env file in both the root folder and the ./database folder. I'm not sure if this is the right way or not.

Hi @Tony-Thawatchai. A couple of things for us to assess here. 

1. I think we can fix this by changing the value in `env_file` to just `./.env` instead of `./database/.env`
2. Can we check if we really need the docker file for postgres? I'm wondering if we even need that if we're combining all services (database, backend application, etc) in a docker-compose

**Comment by Tony-Thawatchai** [comment]
@gmartinsnull resolved your comment and ran the docker-compose build on my local, and DB is up. 

Please kindly leave me a comment if there is anything to improve.

File: airbnb_project/database/airbnb_db_creator.sql [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    baths INT,
+    beds INT,
+    latitude DECIMAL(8,6), 
+    location VARCHAR(20),
+    longitude DECIMAL(9,6),
+    name VARCHAR(100),
+    person_capacity INT,
+    registration_number VARCHAR(20),
+    room_type VARCHAR(20),
+    title VARCHAR(100)
```
Comment (immangat) [inline airbnb_project/database/airbnb_db_creator.sql:12]: new column `is_bath_shared`, boolean, needs to be added.

File: airbnb_project/database/airbnb_db_creator.sql [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    airbnb_listing_id VARCHAR(50),
+    policy_id INT,
+    policy_result BOOLEAN,
+    result_details VARCHAR(255),
+    result_datetime TIMESTAMP,
+    FOREIGN KEY (airbnb_listing_id) REFERENCES listing(airbnb_listing_id),
+    FOREIGN KEY (policy_id) REFERENCES policy(policy_id)
+);
+
+INSERT INTO listing (airbnb_listing_id, baths, beds, latitude, location, longitude, name, person_capacity, registration_number, room_type, title) VALUES
```
Comment (immangat) [inline airbnb_project/database/airbnb_db_creator.sql:32]: mostly correct, new column is needed and format of room type needs to be changed. 

Some sample data: 

|airbnb_listing_id  |bath_is_shared|baths|baths_text|beds|latitude         |location |longitude          |name                                            |person_capacity|registration_number|room_type   |title                 |
|-------------------|--------------|-----|----------|----|-----------------|---------|-------------------|------------------------------------------------|---------------|-------------------|------------|----------------------|
|1161822129037475555|False         |2    |2 baths   |3   |49.25234553007443|Vancouver|-123.22965155214942|Wesbrook Apartment close to UBC                 |6              |Exempt             |entire_home |Apartment in Vancouver|
|976273126562399664 |False         |1.5  |1.5 baths |1   |49.24029         |Vancouver|-123.19577         |Private Room w/Queen Bed in VancouverWest Dunbar|2              |24-160007          |private_room|Room in Vancouver     |

File: airbnb_project/database/airbnb_db_creator.sql [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+CREATE TABLE IF NOT EXISTS listing_policy_result (
+    result_id SERIAL PRIMARY KEY,
+    airbnb_listing_id VARCHAR(50),
+    policy_id INT,
+    policy_result BOOLEAN,
+    result_details VARCHAR(255),
+    result_datetime TIMESTAMP,
+    FOREIGN KEY (airbnb_listing_id) REFERENCES listing(airbnb_listing_id),
+    FOREIGN KEY (policy_id) REFERENCES policy(policy_id)
```
Comment (nam-m) [inline airbnb_project/database/airbnb_db_creator.sql:29]: I suggest adding `ON DELETE CASCADE` at the end of these two so when a listing or policy is deleted, the result would be deleted as well.

File: airbnb_project/airbnb_project/settings_test.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+https://docs.djangoproject.com/en/5.0/ref/settings/
+"""
+
+from .settings import *
+
+# Ensure this setting to indicate it's a test environment
+DEBUG = False  # Set to True or False as per your testing needs
+
+# Ensure this setting to indicate it's a test environment
+DEBUG = False  # Set to True or False as per your testing needs
```
Comment (gmartinsnull) [inline airbnb_project/airbnb_project/settings_test.py:19]: duplicate maybe?

File: airbnb_project/docker-compose.yml [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    healthcheck:
+      test: ["CMD-SHELL", "pg_isready -U ${POSTGRES_USER}"]
+      interval: 30s
+      timeout: 10s
+      retries: 5
+volumes:
+  pgdata:
+    driver: local
+networks:
+  mynetwork:
```
Comment (gmartinsnull) [inline airbnb_project/docker-compose.yml:28]: I'd suggest changing this to something more specific like `airbnb_network`?

**Comment by gmartinsnull** [comment]
A couple of things here:
1. We need to make sure the module `policies` and other new files belong in this PR
2. Add `.DS_Store` to `.gitignore`

File: airbnb_project/docker-compose.yml [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+    container_name: postgres_db
+    env_file:
+      - ./.env
+    environment:
+      POSTGRES_USER: ${POSTGRES_USER}
+      POSTGRES_PASSWORD: ${POSTGRES_PASSWORD}
+      POSTGRES_DB: ${POSTGRES_DB}
+    volumes:
+      - pgdata:/var/lib/postgresql/data
+      - ./database/airbnb_db_creator.sql:/docker-entrypoint-initdb.d/airbnb_db_creator.sql
```
Comment (nam-m) [inline airbnb_project/docker-compose.yml:14]: I think there's no need to initialize db with `./database/airbnb_db_creator.sql`, so can be removed in the future. In django, this can be done with `python manage.py makemigrations`, then `python manage.py migrate`

**Review by nam-m — COMMENTED** [review]
I think we should use a fixed version `postgres` for stability.

**Comment by Tony-Thawatchai** [comment]
@gmartinsnull  @nam-m 

I have updated this PR according to your comment. Please kindly review. 

- make change on airbnb_project/docker-compose.yml
- remove ./database/airbnb_db_creator.sql as per Nam's comment
- respond to other comment

File: airbnb_project/database/.DS_Store
Comment (gmartinsnull) [inline airbnb_project/database/.DS_Store:1]: I noticed you added this to `.gitignore` but I think it's been already committed? I'd suggest we delete and do another commit.

**Comment by gmartinsnull** [comment]
Hi @Tony-Thawatchai, good work on the changes! Just one thing. I wasn't able to find it, have we removed `Dockerfile`?

**Review by nam-m — COMMENTED** [review]
I'm good with your changes, just make sure to have your branch up to date with main

**Comment by Tony-Thawatchai** [comment]
@nam-m  I have updated this PR, resolving all comments. Please kindly review. 

@immangat

**Review by nam-m — APPROVED** [review]
Looks good to me. Thank you
